idx/metric_definition.go

```diff
@@ -0,0 +1,546 @@
+	Key   uintptr
+	Value uintptr
+}
+
+// String returns a Key/Value pair in the form of
+// 'key=value'
+func (t *TagKeyValue) String() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func (t TagKeyValue) createTagKeyValue() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func parseTagKeyValue(tag string) TagKeyValue {
+	var tkv TagKeyValue
+	if strings.Contains(tag, ";") {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	eqPos := strings.Index(tag, "=")
+	if eqPos < 0 {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Key = key
+
+	value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Value = value
+	return tkv
+}
+
+type KeyValuesSlice []TagKeyValue
+
+func (kvs KeyValuesSlice) Len() int      { return len(kvs) }
+func (kvs KeyValuesSlice) Swap(i, j int) { kvs[i], kvs[j] = kvs[j], kvs[i] }
+func (kvs KeyValuesSlice) Less(i, j int) bool {
+	k1, _ := IdxIntern.GetStringFromPtr(kvs[i].Key)
+	k2, _ := IdxIntern.GetStringFromPtr(kvs[j].Key)
+	return k1 < k2
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues struct {
+	KeyValues KeyValuesSlice
+}
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t *TagKeyValues) Strings() []string {
+	tags := make([]string, len((*t).KeyValues))
+	for i, tag := range (*t).KeyValues {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+func (t *TagKeyValues) ExtensionType() int8 {
+	return 90
+}
+
+func (t *TagKeyValues) Len() int {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		if kv == "=" || kv == "" {
```

Comment:
does this mean we added a tag with zero-length key and value? that seems illegal

Reply:
Yes, I think I just forgot to remove that after I fixed some earlier bugs in the code.